File: packages/types/src/plan.ts
import { z } from "zod";

/**
 * Plans. There is exactly one today — everyone is on Free — but the plan is a
 * real, persisted attribute (master migration v3, `tenants.plan`, default
 * 'free') so a paid tier later is a data change, not a schema change. The
 * feature copy lives HERE so the marketing pricing card, Settings → Account and
 * the native apps all render from one list and can't drift.
 *
 * Honest footnote every surface carries: fair-use abuse caps still exist
 * (`enforceQuota` daily caps, the live server's per-device push cap) — "unlimited"
 * means "no plan limit", not "no abuse limit".
 */
export const planIdSchema = z.enum(["free"]);
export type PlanId = z.infer<typeof planIdSchema>;

export interface PlanFeature {
  key: string;
  label: string;
}

export interface PlanDefinition {
  name: string;
  /** USD per month. */
  price: number;
  features: readonly PlanFeature[];
}

export const PLAN_FEATURES = {
  free: {
    name: "Free",
    price: 0,
    features: [
      { key: "bookmarks", label: "Unlimited bookmarks" },
      { key: "live-sessions", label: "Unlimited live sessions" },
      { key: "ai-credits", label: "1,000 AI chat credits every week" },
      { key: "byok", label: "Bring your own key — unmetered AI on your provider" },
    ],
  },
} as const satisfies Record<PlanId, PlanDefinition>;

/** The small print under any pricing/plan card. */
export const PLAN_FAIR_USE_NOTE = "Fair-use limits apply";

/** GET /api/account → the caller's plan. `"free"` in single-tenant/open mode. */
export const accountResponseSchema = z.object({
  plan: planIdSchema,
});
export type AccountResponse = z.infer<typeof accountResponseSchema>;

File: apps/macos/Sources/Models/Plan.swift
import Foundation

/// One line on the plan card — mirror of `PLAN_FEATURES.free.features` in
/// `packages/types/src/plan.ts`. The web renders straight from that constant;
/// this app can't import TypeScript, so the copy is duplicated here VERBATIM.
/// If the list changes there, change it here in the same commit.
struct PlanFeature: Hashable, Identifiable, Sendable {
    let key: String
    let label: String
    var id: String { key }
}

/// `GET /api/account` → `{plan}` (`accountResponseSchema`). Every field optional
/// so the route can grow — and an older server's empty object still decodes.
struct AccountResponse: Codable, Sendable, Equatable {
    var plan: String?
}

/// Everyone is on Free today. `resolve` maps whatever `GET /api/account`
/// reports (or fails to report) onto the one plan this app knows how to draw.
struct PlanInfo: Hashable, Sendable {
    let id: String
    let name: String
    /// USD per month. Free is 0; formatted by the UI.
    let price: Int
    let features: [PlanFeature]

    static let free = PlanInfo(
        id: "free",
        name: "Free",
        price: 0,
        features: [
            PlanFeature(key: "bookmarks", label: "Unlimited bookmarks"),
            PlanFeature(key: "live-sessions", label: "Unlimited live sessions"),
            PlanFeature(key: "ai-credits", label: "2,000 AI chat credits every week"),
            PlanFeature(key: "byok", label: "Bring your own key — unmetered AI on your provider"),
        ]
    )

    /// Unknown/absent plan ids fall back to Free — the only plan that exists,
    /// and what the server reports in single-tenant/open mode.
    static func resolve(_ planId: String?) -> PlanInfo {
        switch planId {
        case "free", nil: return .free
        default: return .free
        }
    }

    /// "Free plan" — the badge text.
    var badgeTitle: String { "\(name) plan" }
}
